crates/web5/src/apid/credentials/verifiable_credential_1_1.rs

```diff
@@ -98,26 +171,302 @@ impl VerifiableCredential {
         }
     }
 
-    pub fn sign(&self, _signer: Arc<dyn Signer>) -> Result<String> {
-        println!("VerifiableCredential.sign() called");
-        Ok(String::default())
+    pub fn sign(&self, bearer_did: &BearerDid) -> Result<String> {
+        // default to first VM
+        let key_id = bearer_did.document.verification_method[0].id.clone();
+        let signer = bearer_did.get_signer(key_id.clone())?;
+
+        self.sign_with_signer(&key_id, signer)
+    }
+
+    pub fn sign_with_signer(&self, key_id: &str, signer: Arc<dyn Signer>) -> Result<String> {
+        let mut payload = JwtPayload::new();
+        let vc_value = serde_json::to_value(self)?;
+        payload.set_claim("vc", Some(vc_value))?;
+        payload.set_issuer(&self.issuer.to_string());
+        payload.set_jwt_id(&self.id);
+        payload.set_subject(&self.credential_subject.id);
+        payload.set_not_before(&self.issuance_date);
+        payload.set_issued_at(&SystemTime::now());
+        if let Some(exp) = &self.expiration_date {
+            payload.set_expires_at(exp)
+        }
+
+        let jose_signer = JoseSigner {
+            kid: key_id.to_string(),
+            signer,
+        };
+
+        let mut header = JwsHeader::new();
+        header.set_token_type("JWT");
+        let vc_jwt = josekit::jwt::encode_with_signer(&payload, &header, &jose_signer)?;
+
+        Ok(vc_jwt)
```

Comment:
I wonder if we should validate/verify before we give the Ok, I wonder if invalid values can sneak through to this point and we give a VC that is signed with invalid properties

Reply:
hmmmmmm maybe, but I'm not convinced yet